src/App.tsx

```diff
@@ -1,223 +1,172 @@
+import { deleteTodoById, getTodos, updateTodoById } from './api/todos';
 import { AuthContext } from './components/Auth/AuthContext';
+import { Todo } from './types/Todo';
+import { CompletedFilter } from './types/CompletedFilter';
+import { Header } from './components/Header';
+import { Todolist } from './components/Todolist';
+import { Footer } from './components/Footer';
+import { ErrorNotification } from './components/ErrorNotification';
+import { filteredTodos } from './helper/Helper';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [completedFilter, setCompletedFilter] = useState(CompletedFilter.All);
+  const [temporaryTodo, setTemporaryTodo] = useState<Todo | null>(null);
+  const [deletingTodosIds, setDeletingTodosIds] = useState([0]);
+  const [updatingTodosIds, setUpdatingTodosIds] = useState([0]);
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
-  const newTodoField = useRef<HTMLInputElement>(null);
+
+  const showError = useCallback((message: string) => {
+    setErrorMessage(message);
+
+    setTimeout(() => setErrorMessage(''), 3000);
+  }, []);
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    if (user) {
+      getTodos(user.id)
+        .then(setTodos)
+        .catch(() => showError('Can\'t load todos'));
     }
   }, []);
 
+  const closeErrorMessage = useCallback(() => {
+    setErrorMessage('');
+  }, []);
+
+  const visibleTodos = useMemo(() => {
+    return filteredTodos(todos, completedFilter);
+  }, [todos, completedFilter]);
+
+  const updateTodo = (todoToChange: Todo) => {
+    setUpdatingTodosIds(
+      (prev) => [...prev, todoToChange.id || 0],
+    );
+
+    updateTodoById(todoToChange.id || 0, {
+      title: todoToChange.title,
+      completed: todoToChange.completed,
+    }).then(() => {
+      setTodos(currentTodos => currentTodos.map(todo => {
+        if (todo.id === todoToChange.id) {
+          return todoToChange;
+        }
+
+        return todo;
+      }));
+
+      setUpdatingTodosIds((currentUpdatingTodos) => (
+        currentUpdatingTodos
+          .filter(updatingId => updatingId !== todoToChange.id)
+      ));
+    }).catch(() => showError('Unable to update a todo'));
+  };
+
+  const isAllTodosCompleted = todos.every(todo => todo.completed);
+
+  const toggleAll = () => {
+    if (!isAllTodosCompleted) {
+      todos.forEach(todo => {
+        if (!todo.completed) {
+          updateTodo({
+            ...todo,
+            completed: true,
+          });
+        }
+      });
+    } else {
+      todos.forEach(todo => {
+        if (todo.completed) {
+          updateTodo({
+            ...todo,
+            completed: false,
+          });
+        }
+      });
+    }
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
```

Comment:
You are passing this callback to another component, so you should wrap it with useCallback